File: 1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.py
class Solution:
    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
        def backtrack(i: int, remain: Dict[int, int]) -> int:
            if i == n:
                return 0
            
            total_score, used = 0, True
            word = Counter(words[i])
            for c, freq in word.items():
                if freq > remain[c]:
                    used = False
                    break
            
            if used:
                for c, freq  in word.items():
                    remain[c] -= freq
                    total_score += freq * score[ord(c) - 97]
                    
                total_score += backtrack(i + 1, remain)
                
                for c, freq in word.items():
                    remain[c] += freq
                
                total_score = max(total_score, backtrack(i + 1, remain))
                
            else:
                total_score = backtrack(i + 1, remain)
                
            return total_score
        
        
        n, remain = len(words), Counter(letters)
            
        return backtrack(0, remain)

File: 1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.cpp
class Solution {
private:
    int n;
    
    int backtrack(int i, unordered_map<char, int>& remain, vector<string>& words, vector<int>& score){
        if (i == n)
            return 0;
        
        int totalScore = 0;
        bool used = true;
        unordered_map<char, int> word;
        for (char& c:words[i]) ++word[c];
        
        for (auto& [c, freq] : word)
            if (freq > remain[c]){
                used = false;
                break;
            }
        
        if (used){
            for (auto& [c, freq] : word){
                remain[c] -= freq;
                totalScore += freq * score[c - 97];
            }
            totalScore += backtrack(i + 1, remain, words, score);
            
            for (auto& [c, freq] : word)
                remain[c] += freq;
            
            totalScore = max(totalScore, backtrack(i + 1, remain, words, score));
        }
        else
            totalScore = backtrack(i + 1, remain, words, score);
            
        return totalScore;
    }
    
public:
    int maxScoreWords(vector<string>& words, vector<char>& letters, vector<int>& score) {
        this -> n = words.size();
        unordered_map<char, int> remain;
        for (char& letter : letters)
            ++remain[letter];
        
        return backtrack(0, remain, words, score);
    }
};